staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -1219,6 +1222,190 @@ func RunSendInitialEventsBackwardCompatibility(ctx context.Context, t *testing.T
+					ObjectMeta: metav1.ObjectMeta{
+						ResourceVersion: "3",
+						Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+					},
+				}},
+			},
+		},
+		{
+			name:                               "allowWatchBookmarks=true, sendInitialEvents=true, RV=1",
+			allowWatchBookmarks:                true,
+			sendInitialEvents:                  &trueVal,
+			resourceVersion:                    "1",
+			initialPods:                        []*example.Pod{makePod("2"), makePod("3")},
+			expectedInitialEventsInRandomOrder: []watch.Event{{Type: watch.Added, Object: makePodWithRV(2)}, {Type: watch.Added, Object: makePodWithRV(3)}},
+			expectedInitialEventsInStrictOrder: []watch.Event{
+				{Type: watch.Bookmark, Object: &example.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						ResourceVersion: "3",
+						Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+					},
+				}},
+			},
+		},
+		{
+			name:                                 "allowWatchBookmarks=false, sendInitialEvents=true, RV=unset",
+			sendInitialEvents:                    &trueVal,
+			initialPods:                          []*example.Pod{makePod("2")},
+			expectedInitialEventsInRandomOrder:   []watch.Event{{Type: watch.Added, Object: makePodWithRV(2)}},
+			podsAfterEstablishingWatch:           []*example.Pod{makePod("3")},
+			expectedEventsAfterEstablishingWatch: []watch.Event{{Type: watch.Added, Object: makePodWithRV(3)}},
+		},
+		{
+			name:                               "allowWatchBookmarks=false, sendInitialEvents=true, RV=0",
+			sendInitialEvents:                  &trueVal,
+			resourceVersion:                    "0",
+			initialPods:                        []*example.Pod{makePod("2"), makePod("3")},
+			expectedInitialEventsInRandomOrder: []watch.Event{{Type: watch.Added, Object: makePodWithRV(2)}, {Type: watch.Added, Object: makePodWithRV(3)}},
+		},
+		{
+			name:              "allowWatchBookmarks=false, sendInitialEvents=true, RV=1",
+			sendInitialEvents: &trueVal,
+			resourceVersion:   "1",
+			initialPods:       []*example.Pod{makePod("2"), makePod("3")},
+			// make sure we only get initial events that are > initial RV (1)
+			expectedInitialEventsInRandomOrder: []watch.Event{{Type: watch.Added, Object: makePodWithRV(2)}, {Type: watch.Added, Object: makePodWithRV(3)}},
+		},
+		{
+			name:                                 "sendInitialEvents=false, RV=unset",
+			sendInitialEvents:                    &falseVal,
+			initialPods:                          []*example.Pod{makePod("2"), makePod("3")},
+			podsAfterEstablishingWatch:           []*example.Pod{makePod("4")},
+			expectedInitialEventsInRandomOrder:   []watch.Event{{Type: watch.Added, Object: makePodWithRV(2)}, {Type: watch.Added, Object: makePodWithRV(3)}},
+			expectedEventsAfterEstablishingWatch: []watch.Event{{Type: watch.Added, Object: makePodWithRV(4)}},
+		},
+
+		// the following case works only against the etcd implementation
```

Comment:
maybe we could change the semantic for the etcd impl to start at the "MostRecent" in this case.